Multi Series Dashcards Render with Filters Applied To All Series

adam-james-v opened the pull request:
Fixes: #39083

Multi-series dashcards now properly render when a subscription is set up with filters that impact the values of each card in the multi-series card.

![image](https://github.com/metabase/metabase/assets/21064735/ef79e0b2-3289-4dd2-98e5-927aa45a8e19)

adam-james-v committed:
Multi Series Dashcards Render with Filters Applied To All Series

Fixes: #39083

Multi-series dashcards now properly render when a subscription is set up with filters that impact the values of each
card in the multi-series card.

adam-james-v committed:
Small refactor for executing dashcards in a pulse context.

- move execute-dashboard-subscription to pulse.util
- this fn now will also execute all card series cards
  and put those results into a list on the key :series-results in the dashcard
  which is then used in the render multimethod to send the correct data to the javascript context

A small todo was taken care of as well, we can now use :dashboard-subscription in the :context key

adam-james-v committed:
Fix up render fn

The shape of hte data going into the javascript was a bit unclear before. This keeps things a little more
understandable.

Next step is probably to create some Malli schemas inside the js_svg namespace to document and enforce the needed data shapes

adam-james-v committed:
pull :data up a level from :result in the series cards

adam-james-v committed:
Merge branch 'master' into bugfix-39083-multi-question-cards-filter-incorrectly

adam-james-v committed:
Merge branch 'master' into bugfix-39083-multi-question-cards-filter-incorrectly

qnkhuat commented on src/metabase/pulse/util.clj:
```diff
@@ -45,24 +46,49 @@
+
+  This function should be executed under pulse's creator permissions."
+  [dashcard parameters]
+  (let [{card-id      :card_id
+         dashboard-id :dashboard_id} dashcard]
```
I think we could pack this these let with the inner let

qnkhuat commented on src/metabase/pulse/render/body.clj:
```diff
@@ -528,37 +527,35 @@
 (mu/defmethod render :javascript_visualization :- formatter/RenderedPulseCard
   [_chart-type render-type _timezone-id card dashcard data]
-  (let [combined-cards-results (if dashcard
-                                 (pu/execute-multi-card card dashcard)
```
awesome, I love that you moved the execution up the stack. We basically executed a dashcard twice before! so this will improve the performance somewhat.

qnkhuat commented on src/metabase/pulse/util.clj:
```diff
@@ -45,24 +46,49 @@
+                                          card
+                                          (t2/select-one :model/Card :id card-id))
+                              :dashcard dashcard
+                              :type     :card
```
I'd not die on this hill, but what about introducing a new type of `multi-cards` so we could differentiate between a single card and a card that has multi-series results?

I don't like the fact that we're assuming a card and a multi-card is the samething.

adam-james-v replied:
Yea, I've been thinking about that too. I'd like to merge this PR first, though.

I'm also working on multi-series cards csv/xlsx downloads/subscription attachments, where it will probably be much better to have mulit-cards as their own thing, like you suggest. 

So, for this PR, I think no, but in a next step, yea, almost certainly a good idea.

qnkhuat commented on src/metabase/pulse/render/body.clj:
```diff
@@ -528,37 +527,35 @@
-        {rendered-type :type content :content} (js-svg/javascript-visualization cards-with-data dashcard-viz-settings)]
+  (let [series-cards-results                   (:series-results dashcard)
+        cards-with-data                        (->> series-cards-results
+                                                    (map result-k->data-k)
```
```suggestion
                                                    (map #(clojure.set/rename-keys % {:result :data}))
```
this should do the same thing

adam-james-v replied:
Ah, turns out I do need to use a function here to achieve what I need: The :data key inside the result key of the card passed in is what we need, we don't want to just rename :result to :data because that results in this:
`{:data {:data {...}}` when what we actually want is: `{:data {...}}`

adam-james-v committed:
Update src/metabase/pulse/render/body.clj

Co-authored-by: Ngoc Khuat <qn.khuat@gmail.com>

adam-james-v committed:
Remove unneeded let

adam-james-v committed:
set/rename-keys won't work here because we raise the :data key

I've realised a suggested change to just use set/rename-keys won't do what we need. We need instead to raise :data
from within the result key up to the same level (and dissoc result).

I've renamed the helper function to make its purpose a bit clearer

adam-james-v committed:
Merge branch 'master' into bugfix-39083-multi-question-cards-filter-incorrectly

qnkhuat reviewed (approved):
nice!
